File: src/main/java/pl/training/workshop/m6/s11_safecomposite/step2/Combo.java
package pl.training.workshop.m6.s11_safecomposite.step2;

import java.util.List;

import pl.training.workshop.shared.Money;

/** Krok 2: węzeł z dziećmi podanymi przy tworzeniu - add() nie istnieje nigdzie. */
public record Combo(String name, List<MenuComponent> children) implements MenuComponent {
    public Combo {
        children = List.copyOf(children);
    }

    public static Combo of(String name, MenuComponent... children) {
        return new Combo(name, List.of(children));
    }

    @Override
    public Money price() {
        Money total = Money.ZERO;
        for (MenuComponent child : children) {
            total = total.plus(child.price());
        }
        return total;
    }

    @Override
    public String describe() {
        String text = name + " " + price();
        if (children.isEmpty()) {
            return text;
        }
        return text + " " + children.stream().map(MenuComponent::describe).toList();
    }
}

File: csharp/src/Training.Workshop/M6/S11SafeComposite/Step2/Combo.cs
using Training.Workshop.Shared;

namespace Training.Workshop.M6.S11SafeComposite.Step2;

/// <summary>Krok 2: węzeł z dziećmi podanymi przy tworzeniu - Add() nie istnieje nigdzie.</summary>
public sealed record Combo : MenuComponent
{
    public Combo(string name, IEnumerable<MenuComponent> children)
    {
        Name = name;
        Children = children.ToList().AsReadOnly();
    }

    public override string Name { get; }

    public IReadOnlyList<MenuComponent> Children { get; }

    public static Combo Of(string name, params MenuComponent[] children)
    {
        return new Combo(name, children);
    }

    public override Money Price
    {
        get
        {
            var total = Money.Zero;
            foreach (var child in Children)
            {
                total = total.Plus(child.Price);
            }
            return total;
        }
    }

    public override string Describe()
    {
        var text = Name + " " + Price;
        if (Children.Count == 0)
        {
            return text;
        }
        return text + " [" + string.Join(", ", Children.Select(child => child.Describe())) + "]";
    }
}
